File: src/engine.js
// =============================================================================
// Arcade — engine: the game contract, registry, and deterministic runtime
// =============================================================================
// A game is pure data + pure functions:
//   setup(ctx) → state           build the initial state from players + rng
//   reducers[action](state, payload, ctx) → nextState   one move
//   view(state, playerId) → playerView   what a player may see (hides secrets)
//   isOver / result               end + outcome
//
// Because the runtime owns a SEEDED rng and applies actions in order, every
// device that creates a match with the same seed and replays the same action
// log reaches byte-identical state — that's lockstep multiplayer with no
// server (see ./transport.js). "Add your own game" = call defineGame() and
// register it; people can clone the repo and drop a game in games/.
// =============================================================================

import { makeRng } from './rng.js';

/**
 * Define a game. Returns a frozen, validated descriptor.
 * @param {object} def
 */
export function defineGame(def) {
  for (const k of ['id', 'setup', 'reducers']) {
    if (!def[k]) throw new Error(`defineGame: "${k}" is required`);
  }
  return Object.freeze({
    id: def.id,
    title: def.title ?? def.id,
    minPlayers: def.minPlayers ?? 1,
    maxPlayers: def.maxPlayers ?? 8,
    setup: def.setup,
    reducers: Object.freeze({ ...def.reducers }),
    view: def.view ?? ((state) => state),
    isOver: def.isOver ?? (() => false),
    result: def.result ?? (() => null),
  });
}

/** Registry of games. The default export `games` is a shared instance. */
export class GameRegistry {
  constructor() {
    this._byId = new Map();
  }
  register(game) {
    if (this._byId.has(game.id)) throw new Error(`game already registered: ${game.id}`);
    this._byId.set(game.id, game);
    return game;
  }
  get(id) {
    return this._byId.get(id) ?? null;
  }
  has(id) {
    return this._byId.has(id);
  }
  list() {
    return [...this._byId.values()];
  }
}

export const games = new GameRegistry();

/**
 * Create a match. Deterministic: same (game, players, seed, options) + same
 * dispatched actions ⇒ identical state on every device.
 *
 * @param {object} opts
 * @param {object} opts.game     a defineGame() descriptor
 * @param {string[]} opts.players ordered player ids
 * @param {string|number} opts.seed
 * @param {object} [opts.options]
 */
export function createMatch({ game, players, seed, options = {} }) {
  if (!game) throw new Error('createMatch: game is required');
  if (!Array.isArray(players) || players.length < game.minPlayers) {
    throw new Error(`${game.id}: needs at least ${game.minPlayers} player(s)`);
  }
  if (players.length > game.maxPlayers) {
    throw new Error(`${game.id}: at most ${game.maxPlayers} player(s)`);
  }

  const rng = makeRng(seed);
  let state = game.setup({ players, rng, options });
  const log = [];

  const match = {
    game,
    players,
    seed,
    get state() {
      return state;
    },
    get log() {
      return log.slice();
    },
    view: (playerId) => game.view(state, playerId),
    isOver: () => game.isOver(state),
    result: () => game.result(state),

    /**
     * Apply an action. Records it to the log for replay / transport.
     * @param {{type:string, payload?:any, playerId?:string}} action
     */
    dispatch(action) {
      if (game.isOver(state)) throw new Error(`${game.id}: match is over`);
      const reducer = game.reducers[action.type];
      if (!reducer) throw new Error(`${game.id}: unknown action "${action.type}"`);
      state = reducer(state, action.payload, { rng, playerId: action.playerId, players });
      log.push({ type: action.type, payload: action.payload, playerId: action.playerId });
      return state;
    },
  };
  return match;
}

/**
 * Rebuild a match's state by replaying an action log (e.g. a late joiner, or a
 * peer verifying another's state). Same seed + same log ⇒ same state.
 */
export function replayMatch({ game, players, seed, options = {}, log = [] }) {
  const match = createMatch({ game, players, seed, options });
  for (const action of log) match.dispatch(action);
  return match;
}

File: go/engine/engine.go
// Package engine is Arcade's deterministic game runtime in Go — the sibling of
// the JS engine. A game is a pure Game[S]; a Match[S] owns a seeded rng and
// applies actions in order, so the same seed + same action log reproduces the
// same state (and matches the JS engine when the game logic matches).
package engine

import "github.com/ericdequ/Arcade/go/rng"

// Action is one move.
type Action struct {
	Type     string         `json:"type"`
	Payload  map[string]any `json:"payload,omitempty"`
	PlayerID string         `json:"playerId,omitempty"`
}

// Ctx is passed to a game's Reduce.
type Ctx struct {
	RNG      *rng.RNG
	PlayerID string
	Players  []string
}

// Game is a pure, deterministic game over a state type S.
type Game[S any] interface {
	ID() string
	Setup(players []string, r *rng.RNG, options map[string]any) S
	Reduce(state S, action Action, ctx Ctx) S
	IsOver(state S) bool
}

// Match runs a Game deterministically.
type Match[S any] struct {
	game    Game[S]
	players []string
	rng     *rng.RNG
	state   S
	log     []Action
}

// NewMatch builds a match. Same (game, players, seed, options) + same dispatched
// actions ⇒ identical state on every device, in any language.
func NewMatch[S any](g Game[S], players []string, seed string, options map[string]any) *Match[S] {
	r := rng.New(seed)
	return &Match[S]{game: g, players: players, rng: r, state: g.Setup(players, r, options)}
}

// State returns the current full state.
func (m *Match[S]) State() S { return m.state }

// Players returns the ordered player ids.
func (m *Match[S]) Players() []string { return m.players }

// IsOver reports whether the game has ended.
func (m *Match[S]) IsOver() bool { return m.game.IsOver(m.state) }

// Log returns a copy of the applied action log (for replay / transport).
func (m *Match[S]) Log() []Action { return append([]Action(nil), m.log...) }

// Dispatch applies an action and records it.
func (m *Match[S]) Dispatch(a Action) S {
	if m.game.IsOver(m.state) {
		return m.state
	}
	m.state = m.game.Reduce(m.state, a, Ctx{RNG: m.rng, PlayerID: a.PlayerID, Players: m.players})
	m.log = append(m.log, a)
	return m.state
}

// Replay rebuilds a match by re-applying an action log.
func Replay[S any](g Game[S], players []string, seed string, options map[string]any, log []Action) *Match[S] {
	m := NewMatch(g, players, seed, options)
	for _, a := range log {
		m.Dispatch(a)
	}
	return m
}
